### PR TITLE
txn: Optimize pessimistic transaction by supporting locking with conflict

### LINKED ISSUE #40537: Support aggressive-locking model for pessimistic transaction
This is a corresponding issue of: https://github.com/tikv/tikv/issues/13298

Make use of the aggressive locking model provided in client-go and TiKV. Support enabling it with a system variable.

- Design (TiDB part): https://github.com/pingcap/tidb/pull/37518
- PR: https://github.com/pingcap/tidb/pull/35588

### PR DESCRIPTION
**Pull request by MyonKeminta**
### What problem does this PR solve?

Issue Number: Close #40537

Problem Summary: 

### What is changed and how it works?

Requires https://github.com/tikv/client-go/pull/651. 

This PR optimizes pessimistic transaction by supporting pessimistic-locking with conflict and key-grained locking and waiting in TiKV. A system variable controlling the optimization is introduced.

### Check List

Tests 

- [x] Unit test
- [x] Integration test
- [x] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [x] Performance regression: Consumes more CPU
- [x] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [x] Affects user behaviors
- [ ] Contains syntax changes
- [x] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
Support aggressive locking model for pessimistic transaction, which can reduce tail latency of point-get DMLs in pessimistic transactions that have frequent lock contention. It can be enabled by the system variable `tidb_pessimistic_txn_aggressive_locking`.
```

### CODE REVIEW

File: sessionctx/context.go [hunk 1]
```diff
@@ -135,7 +135,8 @@ type Context interface {
 
 	// StmtCommit flush all changes by the statement to the underlying transaction.
 	StmtCommit()
-	// StmtRollback provides statement level rollback.
+	// StmtRollback provides statement level rollback. The parameter `forPessimisticRetry` should be true iff it's used
```
Comment (sticnarf) [inline sessionctx/context.go:138]: Where is the `forPessimisticRetry` parameter?

File: sessionctx/variable/tidb_vars.go [hunk 2]
```diff
@@ -789,6 +789,10 @@ const (
 
 	// TiDBStoreBatchSize indicates the batch size of coprocessor in the same store.
 	TiDBStoreBatchSize = "tidb_store_batch_size"
+
+	// TiDBPessimisticTransactionAggressiveLocking controls whether aggressive locking for pessimistic transaction
+	// is enabled.
+	TiDBPessimisticTransactionAggressiveLocking = "tidb_pessimistic_transaction_aggressive_locking"
```
Comment (sticnarf) [inline sessionctx/variable/tidb_vars.go:795]: Most TiDB system variables use `txn` to represent "transaction". I'm not sure if it's better to follow the convention...

File: sessionctx/context.go (hunk 1 above)
Comment (MyonKeminta) [reply sessionctx/context.go:138]: Ah sorry, it was removed and I forgot to change back the comment

File: store/driver/txn/txn_driver.go [hunk 3]
```diff
@@ -72,8 +73,38 @@ func (txn *tikvTxn) CacheTableInfo(id int64, info *model.TableInfo) {
...
 func (txn *tikvTxn) LockKeys(ctx context.Context, lockCtx *kv.LockCtx, keysInput ...kv.Key) error {
 	keys := toTiKVKeys(keysInput)
+	if len(keys) > 1 && txn.IsInAggressiveLockingMode() {
+		// Only allow aggressive locking if it only needs to lock one key. Considering that it's possible that a
+		// statement causes multiple calls to `LockKeys` (which means some keys may have been locked in aggressive
+		// locking mode), here we exit aggressive locking mode by calling DoneAggressiveLocking instead of cancelling.
+		// Then the previously-locked keys during execution in this statement (if any) will be turned into the state
+		// as if they were locked in normal way.
+		// Note that the issue https://github.com/pingcap/tidb/issues/35682 also exists here.
+		txn.DoneAggressiveLocking(ctx)
```
[the remark this answers is not in the document]
Comment (MyonKeminta) [reply store/driver/txn/txn_driver.go:83]: Yes. It should be cheap when there are actually nothing locked during the aggressive locking.

File: session/txn.go [hunk 4]
```diff
@@ -449,6 +462,67 @@ func (txn *LazyTxn) LockKeys(ctx context.Context, lockCtx *kv.LockCtx, keys ...k
...
 }
 
+// StartAggressiveLocking wraps the inner transaction to support using aggressive locking with lazy initialization.
+func (txn *LazyTxn) StartAggressiveLocking() {
+	if txn.Valid() {
+		txn.Transaction.StartAggressiveLocking()
+	} else if txn.pending() {
+		txn.enterAggressiveLockingOnValid = true
+	} else {
+		panic("trying to start aggressive locking on a transaction in invalid state")
```
[the remark this answers is not in the document]
Comment (ekexium) [reply session/txn.go:480]: ERROR logs might be ignored. We could let it return an error.

**Comment by MyonKeminta** [comment]
...please hold on. Something went wrong after merging master...

File: session/txn.go [hunk 5]
```diff
@@ -621,7 +702,14 @@ func (s *session) StmtCommit() {
...
 
 // StmtRollback implements the sessionctx.Context interface.
-func (s *session) StmtRollback() {
+func (s *session) StmtRollback(ctx context.Context, isForPessimisticRetry bool) {
+	if !isForPessimisticRetry {
+		txnManager := sessiontxn.GetTxnManager(s)
+		err := txnManager.OnStmtRollback(ctx, isForPessimisticRetry)
+		if err != nil {
+			logutil.Logger(ctx).Error("txnManager failed to handle OnStmtRollback", zap.Error(err))
+		}
```
[the remark this answers is not in the document]
Comment (MyonKeminta) [reply session/txn.go:711]: Ah, sorry, it's a mistake during my refactorying

**Comment by MyonKeminta** [comment]
The `LockKeysFunc` introduced in master bypassed some of my logic... fixed......

File: session/txn.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (cfzjywxk) [reply session/txn.go:480]: Yes, adding an ERROR log here and returning an error.

File: session/txn.go [hunk 6]
```diff
@@ -282,6 +289,11 @@ func (txn *LazyTxn) changePendingToValid(ctx context.Context) error {
 	txn.Transaction = t
 	txn.initStmtBuf()
 
+	if txn.enterAggressiveLockingOnValid {
+		txn.enterAggressiveLockingOnValid = false
+		txn.Transaction.StartAggressiveLocking()
```
Comment (cfzjywxk) [inline session/txn.go:294]: We may need telemetry information for example how many transactions use the aggressive locking mode and actually do the aggressive-mode locking. As this is a new feature we could enrich this statistic information early.

File: session/txn.go [hunk 7]
```diff
@@ -456,6 +470,67 @@ func (txn *LazyTxn) LockKeysFunc(ctx context.Context, lockCtx *kv.LockCtx, fn fu
...
+	} else {
+		panic("trying to cancel aggressive locking on a transaction in invalid state")
+	}
+}
+
+// IsInAggressiveLockingMode wraps the inner transaction to support using aggressive locking with lazy initialization.
+func (txn *LazyTxn) IsInAggressiveLockingMode() bool {
+	if txn.Valid() {
+		return txn.Transaction.IsInAggressiveLockingMode()
+	}
```
Comment (cfzjywxk) [inline session/txn.go:527]:
```suggestion
	if txn.Valid() {
	} else if if txn.pending() {
	}
```
looks better we won't miss anything.

Comment (MyonKeminta) [reply session/txn.go:527]: 🤔Actually I think it doesn't matter much since there are early-returns in every branches.

**Review by cfzjywxk — APPROVED** [review]
LGTM

**Comment by MyonKeminta** [comment]
@sticnarf @ekexium PTAL again if you have time, thanks!

**Review by ekexium — APPROVED** [review]

**Review by sticnarf — APPROVED** [review]

**Pull request merged**